velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,283 @@
+        outcome, "failure in S3ReadFile::preadInternal", bucket_, key_);
+
+    result = std::move(outcome).GetResultWithOwnership();
+    auto& stream = result.GetBody();
+    stream.read(reinterpret_cast<char*>(pos), length);
+  }
+  std::shared_ptr<Aws::S3::S3Client> client_;
+  std::string bucket_;
+  std::string key_;
+  int64_t length_ = -1;
+};
+
+namespace filesystems {
+
+namespace S3Config {
+namespace {
+constexpr char const* pathAccessStyle{"hive.s3.path-style-access"};
+constexpr char const* endpoint{"hive.s3.endpoint"};
+constexpr char const* secretKey{"hive.s3.aws-secret-key"};
+constexpr char const* accessKey{"hive.s3.aws-access-key"};
+constexpr char const* sslEnabled{"hive.s3.ssl.enabled"};
+constexpr char const* useInstanceCredentials{
+    "hive.s3.use-instance-credentials"};
+} // namespace
+} // namespace S3Config
+
+// Implement the S3FileSystem
+class S3FileSystem::Impl {
+ public:
+  Impl(const Config* config) : config_(config) {
+    const size_t origCount = initCounter_++;
+    if (origCount == 0) {
+      Aws::SDKOptions awsOptions;
+      awsOptions.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Fatal;
+      Aws::InitAPI(awsOptions);
+    }
+  }
+
+  ~Impl() {
+    const size_t newCount = --initCounter_;
+    if (newCount == 0) {
+      Aws::SDKOptions awsOptions;
+      awsOptions.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Fatal;
+      Aws::ShutdownAPI(awsOptions);
+    }
+  }
+
+  std::string getOptionalProperty(
+      const std::string& name,
+      const std::string& defaultValue) {
+    auto value = config_->get(name);
+    if (!value.hasValue()) {
+      return defaultValue;
+    }
+    return value.value();
+  }
+
+  // Configure default AWS credentials provider chain.
+  std::shared_ptr<Aws::Auth::AWSCredentialsProvider>
+  getDefaultCredentialProvider() {
+    return std::make_shared<Aws::Auth::DefaultAWSCredentialsProviderChain>();
+  }
+
+  // Configure with access and secret keys. Used for on-prem.
+  std::shared_ptr<Aws::Auth::AWSCredentialsProvider>
+  getAccessSecretCredentialProvider(
+      std::string accessKey,
+      std::string secretKey) {
+    return std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+        getAwsString(accessKey), getAwsString(secretKey), getAwsString(""));
+  }
+
+  // Use the input Config parameters and initialize the S3Client
+  void initializeClient() {
+    Aws::Client::ClientConfiguration clientConfig;
+
+    const auto endpoint = getOptionalProperty(S3Config::endpoint, "");
+    clientConfig.endpointOverride = endpoint;
+
+    // default is use SSL
+    const auto useSSL =
+        (getOptionalProperty(S3Config::sslEnabled, "true") == "true");
+    if (useSSL) {
+      clientConfig.scheme = Aws::Http::Scheme::HTTPS;
+    } else {
+      clientConfig.scheme = Aws::Http::Scheme::HTTP;
+    }
+
+    // use virtual addressing for S3 on AWS
+    const auto pathAccessStyle =
+        getOptionalProperty(S3Config::pathAccessStyle, "false");
```

Comment:
Would it be helpful to a template method: getOptionalProperty<bool>(name, defaultValue) ?